### PR TITLE
Fix to not build custom server during `blitz start`

### PR DESCRIPTION
**Pull request by mabadir**
Closes: #2354

### What are the changes and their implications?
- Building Custom Server during Blitz Build
- `startCustomServer` function, now doesn't build by default.
- Introduce a separate `buildCustomServer` fn

## Feature Checklist

- [ ] Integration test added (see [test docs](https://blitzjs.com/docs/contributing#running-tests) if needed)
- [ ] Documentation added/updated (submit PR to [blitzjs.com repo](https://github.com/blitz-js/blitzjs.com))

### CODE REVIEW

File: packages/server/src/next-utils.ts [hunk 1]
```diff
@@ -248,6 +248,7 @@ export function startCustomServer(
   cwd: string,
   config: ServerConfig,
   {watch}: CustomServerOptions = {},
+  buildOnly: boolean = false,
```
Comment (flybayer) [inline packages/server/src/next-utils.ts:251]: Move `buildOnly` into the option object

File: packages/server/src/next-utils.ts [hunk 2]
```diff
@@ -296,7 +297,7 @@ export function startCustomServer(
         "next",
         ...Object.keys(require("blitz/package").dependencies),
         ...Object.keys(pkg.dependencies),
-        ...Object.keys(pkg.devDependencies),
+        // ...Object.keys(pkg.devDependencies),
```
Comment (flybayer) [inline packages/server/src/next-utils.ts:300]: Why?

File: packages/server/src/build.ts [hunk 3]
```diff
@@ -36,9 +36,14 @@ export async function build(config: ServerConfig) {
...
 
   await saveBlitzVersion(buildFolder)
-
   await nextBuild(nextBin, buildFolder, manifest, config)
 
+  try {
+    await startCustomServer(rootFolder, config, {watch, buildOnly: true})
+  } catch (e) {
+    console.log("No custom server found, proceeding")
+  }
```
Comment (flybayer) [inline packages/server/src/build.ts:45]: - I think we should first check if custom server exists instead of doing try/catch

**Review by flybayer — COMMENTED** [review]
Not sure if you are done? Mark as not-draft if you are done.

However, this doesn't fully address the original issue. The custom server is still being built on `blitz start`. I think we should separate the function into `buildCustomServer` and `startCustomServer`.

**Comment by mabadir** [comment]
Yes sorry, I didn't fully understand the main concern from the original issue. I will review this PR today and will push few updates. Sorry about that.

File: packages/server/src/next-utils.ts [hunk 4]
```diff
@@ -317,9 +338,6 @@ export function startCustomServer(
       }
     }
 
-    // eslint-disable-next-line @typescript-eslint/no-floating-promises
-    esbuild.build(esbuildOptions).then(() => {
-      spawnServer()
-    })
+    spawnServer()
```
Comment (flybayer) [inline packages/server/src/next-utils.ts:341]: Looking good! This is a regression for `blitz dev` use case.

Here's the full list. Sorry, I should have put this earlier.

- [x] `blitz build` builds custom server
- [x] `blitz start` runs but does not build custom server
- [ ] `blitz dev` should both build and run custom server (this is working correctly in existing code, but it's not on this PR)

Comment (mabadir) [reply packages/server/src/next-utils.ts:341]: I think adding:
`await buildCustomServer()` to this line will solve it:
https://github.com/blitz-js/blitz/blob/8eb1509b9ad45b647a702381b4f693ed3b8aefe6/packages/server/src/dev.ts#L48

Comment (flybayer) [reply packages/server/src/next-utils.ts:341]: That handles initial build, but not watch. Did you test `blitz dev`?

I think we should conditionally do `spawnServer()` vs 

```
esbuild.build(esbuildOptions).then(() => {
      spawnServer()
})
```

based on watch flag. Currently `esbuildOptions` in this function are not used and so hot reload doesn't work.

Comment (mabadir) [reply packages/server/src/next-utils.ts:341]: You're right, I rushed this one in. This is what I am coming up with now.

```javascript

export function startCustomServer(
    cwd: string,
    config: ServerConfig,
    {watch}: CustomServerOptions = {},
) {
  const serverBuildPath = getCustomServerBuildPath()

  let spawnEnv = getSpawnEnv(config)
  if (config.env === "prod") {
    spawnEnv = {...spawnEnv, NODE_ENV: "production"}
  }

  return new Promise<void>((res, rej) => {
    let process: ChildProcess

    const RESTART_CODE = 777777

    const spawnServer = () => {
      process = spawn("node", [serverBuildPath], {
        cwd,
        env: spawnEnv,
        stdio: "inherit",
      })
          .on("exit", (code: number) => {
            if (code === 0) {
              res()
            } else if (watch && code === RESTART_CODE) {
              spawnServer()
            } else {
              rej(`server.js failed with status code: ${code}`)
            }
          })
          .on("error", (err) => {
            console.error(err)
            rej(err)
          })
    }

    if (watch) {
      // Handle build & Starting server
      const esbuildOptions = getEsbuildOptions()
      esbuildOptions.watch = {
        onRebuild(error) {
          if (error) {
            log.error("Failed to re-build custom server")
          } else {
            log.newline()
            log.progress("Custom server changed - restarting...")
            log.newline()
            //@ts-ignore -- incorrect TS type from node
            process.exitCode = RESTART_CODE
            process.kill("SIGABRT")
          }
        },
      }
      // eslint-disable-next-line @typescript-eslint/no-floating-promises
      esbuild.build(esbuildOptions).then(() => {
        spawnServer()
      })
    } else {
      // No build required, Start server
      spawnServer()
    }
  })
}
```

**Comment by flybayer** [comment]
yay!

**Pull request merged**